### PR TITLE
Update the entityVersion also for downgrades

### LINKED ISSUE #9716: JPA Map storage doesn't downgrade entityVersion when modifying a row written with a future entityVersion
### Describe the bug

When the Map JPA store reads data from a row, it accepts data that is written by another Keycloak version with entityVersion N+1. Thereby it ignores new fields written by a later version during JSON de-serialization. 

When it then writes the data back to the database, it uses its JSON serializer and writes the data in the format of its own JSON schema to the row. Therefore, it should downgrade the entityVersion to N as well, so that is future version of Keycloak can again migrate the data to a later entity version. 

### Version

17.0.0-SNAPSHOT

### Expected behavior

* when data is written to the row, the entity version is always the SUPPORTED_VERSION_* of the implementation

### Actual behavior

* when data is written to the row, the entity version is kept "as is", thereby leaving the N+1 version in the row. This prevents a future migration of the JSON and is inconsistent with the data in the JSON column.

### How to Reproduce?

Manual testing with the database

### Anything else?

There is a discussion on this topic here: https://github.com/keycloak/keycloak/discussions/9640

Possible refactoring: Rework repetitive calls to checkEntityVersionForUpdate() in JpaClientEntity to a method annotated with @PreUpdate and @PrePersist, so that is called whenever a SQL insert or update is triggered. 

This would prevent missing an update to the entity version in cases where Hibernate triggers an update to the database.

### PR DESCRIPTION
**Pull request by ahus1**
as it needs to match the JSON and auxiliary tables

Closes #9716

### CODE REVIEW

**Comment by ahus1** [comment]
As agreed this is a spike/proof of concept for using the `@PreUpdate` annotation to simplify the code in the entities.

I've tested it manually by updating the entityVersion in the database, then updating the entity and only attributes. In both instances the entityVersion is downgraded successfully. 

I'd like to get feedback to this approach before I apply it to more entities. The three remaining calls to updateEntityVersion() are necessary when updating child tables (here: attributes). 

This could be changed to Hibernate listener as shown in #9761 to avoid missing any call to that method. The downside is the additional technical obscurity of Hibernate listeners. While the effect used in #9761 couldn't be achieved in another way, calling this method three times per entity is IMHO tolerable. Therefore 

This needs to be applied to roles as well. To void repeating the JavaDoc of updateEntityVersion(), it could be pulled to a superclass/interface JpaRootEntity that will be introduced in #9761.

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/client/entity/JpaClientEntity.java [hunk 1]
```diff
@@ -129,12 +130,23 @@ public boolean isMetadataInitialized() {
...
+     * of the additional tables might depend on the version.
+     *
+     * Future versions of this method might restrict downgrading to downgrade only from the next version.
      */
-    private void checkEntityVersionForUpdate() {
+    @PreUpdate
+    private void updateEntityVersion() {
         Integer ev = getEntityVersion();
-        if (ev != null && ev < SUPPORTED_VERSION_CLIENT) {
+        if (ev != null && !Objects.equals(ev, SUPPORTED_VERSION_CLIENT)) {
```
Comment (vramik) [inline model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/client/entity/JpaClientEntity.java:149]: Thank you @ahus1. I like the way how it's done.

I think we might restrict downgrading the entity_version only from next keycloak version already. wdyt?
 
+1 for moving the method and javadoc to "future" `JpaRootEntiy`. And also from my POV it's ok to call the method when updating the attributes.

Comment (vramik) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/client/entity/JpaClientEntity.java:149]: It sounds reasonable, +1, thank you!

Comment (sguilhen) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/client/entity/JpaClientEntity.java:149]: +1 to @ahus1 's proposal

**Comment by sguilhen** [comment]
Thanks you @ahus1  - this gets a +1 from me and in my opinion it is perfectly fine to explicitly call `updateEntityVersion` where needed as opposed to having a listener in place. If necessary we can always revisit this decision in the future and refactor all entities accordingly, but for now your proposal looks good to me.

**Comment by hmlnarik** [comment]
Thank you @ahus1 for the investigation!
I am wondering if the approach sketched in https://github.com/keycloak/keycloak/commit/4c0bcee21d106a8451b89ec94c122bf75ac0d8e8 solves an independent issue (object state versioning rather than entity format versioning solved in this PR) and is needed anyway?

**Comment by ahus1** [comment]
@hmlnarik - they solve two independent problems. While this solves the downgrading of the schema problem, the other PR solves a concurrent update of the row problem.

**Comment by ahus1** [comment]
@hmlnarik - update to my previous comment: there's already issue #9874 to track the concurrency problem.

**Comment by ahus1** [comment]
@sguilhen @vramik - finalized my changes and also implemented this for roles. Ready for final review. Build is still running.

**Comment by ahus1** [comment]
reverting to draft status, will rework to use Hibernate API to align implementation with #9894

**Comment by ahus1** [comment]
On hold until #9894 is merged.

**Comment by ahus1** [comment]
This PR is now updated to use the Hibernate SPI to update the entity version. I tested it manually locally and an entity version is downgraded successfully from 2 to 1 even if only an attribute was changed.

I decided to keep the two listeners separate as they serve two different purposes. Please let me know if you want name changes on new methods and classes.

As the listeners are hibernate specific, I moved the two listeners to the hibernate package.

Please let me know what you think, @vramik and @sguilhen. Thanks!

**Comment by ahus1** [comment]
Note: #9944 and #9943 should be merged first, this PR then probably needs a rebase.

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/hibernate/listeners/JpaEntityVersionListener.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+ * Listen on changes on child- and root entities and updates the current entity version of the root.
+ *
+ * This support a multiple level parent-child relationship, where only the upmost parent is locked.
+ */
+public class JpaEntityVersionListener implements PreInsertEventListener, PreDeleteEventListener, PreUpdateEventListener {
+
+    public static final JpaEntityVersionListener INSTANCE = new JpaEntityVersionListener();
+
+    /**
+     * Check if the entity is a child with a parent and force optimistic locking increment on the upmost parent.
```
Comment (sguilhen) [inline model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/hibernate/listeners/JpaEntityVersionListener.java:40]: nitpick: this javadoc was prob copied over from the other listener and doesn't correspond to what the method does.

**Review by sguilhen — COMMENTED** [review]
Thanks @ahus1 , I like what you did here with the listeners. Just left a minor javadoc comment, but overall this looks very good to me.

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/hibernate/listeners/JpaEntityVersionListener.java (hunk 2 above)
Comment (ahus1) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/hibernate/listeners/JpaEntityVersionListener.java]: Thanks, I updated that. The danger of writing comments: they might get out of date. I try to write comment why I do something or the intent, not necessarily how I do things. Think I got a bit off-track here. JavaDoc updated, please re-review.

**Review by sguilhen — APPROVED** [review]

**Review by vramik — APPROVED** [review]
LGTM, thank you @ahus1. I'll rebase https://github.com/keycloak/keycloak/pull/9997 once this one got to be merged.

**Review by hmlnarik — APPROVED** [review]
Approving per @vramik and @sguilhen's reviews

**Pull request merged**